source/base/hdf5.cc

```diff
@@ -105,18 +106,64 @@ namespace HDF5
     , global_no_collective_cause(H5D_MPIO_SET_INDEPENDENT)
   {
     hdf5_reference = std::shared_ptr<hid_t>(new hid_t, [](hid_t *pointer) {
-      // Release the HDF5 resource
-      const herr_t ret = H5Dclose(*pointer);
-      AssertNothrow(ret >= 0, ExcInternalError());
-      (void)ret;
-      delete pointer;
+#  ifdef DEAL_II_WITH_MPI
+      if (std::uncaught_exceptions() > 0)
+        {
+          const unsigned int myid =
+            Utilities::MPI::this_mpi_process(MPI_COMM_WORLD);
+          if (myid == 0)
+            std::cerr
+              << "---------------------------------------------------------\n"
+              << "HDF5 DataSet objects call H5Dclose to end access to the\n"
+              << "dataset and release resources used by it. This call\n"
+              << "requires MPI synchronization. Since an exception is\n"
+              << "currently uncaught, this synchronization will be skipped\n"
+              << "to avoid a possible deadlock. As a result the HDF5 file\n"
+              << "might be corrupted.\n"
+              << "---------------------------------------------------------"
+              << std::endl;
+        }
```

Comment:
I would prefer if you just called `MPI_Abort()` here. The reason is that if you don't, then all of the other processes will hang because this one processor did not participate in the communication. This one processor will then simply go on with its business, will eventually come to a point where it wants to communicate, and hang there because none of the other processors have come to that point. You're really not creating a better outcome, but just postponing the misery.

So, I'd just abort. You then also don't know the paragraph about data corruption in the documentation.